tensorflow/core/kernels/fixed_length_record_reader_op.cc

```diff
@@ -29,73 +32,155 @@ class FixedLengthRecordReader : public ReaderBase {
+
+    // Even if the num_records=0, the footer_bytes_ has to exist:
+    TF_RETURN_IF_ERROR(
+        buffered_inputstream_->ReadNBytes(footer_bytes_, &lookahead_cache_));
 
-    input_buffer_.reset(new io::InputBuffer(file_.get(), kBufferSize));
-    TF_RETURN_IF_ERROR(input_buffer_->SkipNBytes(header_bytes_));
     return Status::OK();
   }
 
   Status OnWorkFinishedLocked() override {
-    input_buffer_.reset(nullptr);
+    buffered_inputstream_.reset(nullptr);
     return Status::OK();
   }
 
   Status ReadLocked(string* key, string* value, bool* produced,
                     bool* at_end) override {
-    // The condition `input_buffer_->Tell() + record_bytes_ > file_pos_limit_`
-    // is to confirm that none of record bytes is out of the range of
-    // file_pos_limit_.
-    // This is necessary for the condition `hop_bytes > 0`. For example.
-    // File: "0123456"
-    // Reader setting: `record_bytes=3`, `hop_bytes=2`, `footer_bytes=0`,
-    //     `header_bytes=0`
-    // Without this checking condition, the forth time the reader will at
-    // this position: "012345|6" and the reading operation will result in
-    // an error.
-    if (input_buffer_->Tell() >= file_pos_limit_ ||
-        input_buffer_->Tell() + record_bytes_ > file_pos_limit_) {
+    // In case hop_bytes_ is in between 0 and record_bytes_,
+    // we will need to hold a cache so that later on we could prefix the cache
+    // with the remaining data to read:
+    // For example, assume record_bytes is 2 and hop_bytes is 1,
+    // and header_bytes is 1,
+    // For file "H123",
+    // We will process the data in the following way:
+    // 1. Read hop_bytes of 1 ("1") and save it in hop cache.
+    // 2. Read record_bytes - hop_bytes = 1 ("2")
+    // 3. Prefix data in 2. with hop cache in one, we have "12",
+    //    this is used for the record.
+    // 4. Shift hop_bytes of 1 from the record, and we put "2" to hop cache.
+    // 5. Continue step 2.
+    // In order to acheive the above, in the following only
+    // 'record_bytes_ - lookahead_cache_.size()' needs to be read. The
+    // lookahead_cache_
+    // is then prefixed to piece together the whole record.
+
+    if (record_number_ == 0) {
+      // we will peek in "record_bytes_ - hop_bytes_"
+      // before we read the first record.
+      // Note lookahead_cache_ should have hold footer_bytes_ in starup
+      if (0 < hop_bytes_ && hop_bytes_ < record_bytes_) {
+        int bytes_to_read = record_bytes_ - hop_bytes_;
+        Status s = buffered_inputstream_->ReadNBytes(bytes_to_read, value);
+        if (!s.ok() && !errors::IsOutOfRange(s)) {
+          return s;
+        }
+
+        // EOF reached
+        if (value->size() == 0) {
+          *at_end = true;
+          return Status::OK();
+        }
+        // Now
+        lookahead_cache_.append(*value, 0, bytes_to_read);
+        value->clear();
+      }
+    }
+
+    int bytes_to_read = record_bytes_;
+    if (0 < hop_bytes_) {
```

Comment:
Thanks @saxenasaurabh for the review. Yes that would be better to understand. The PR has been updated.